### PR TITLE
Add support for vlan update on ovs bridges

### LINKED ISSUE #57018: openvswitch_bridge does not update vlan on existing bridge.
##### SUMMARY

When creating a fake bridge with a parent, this module takes a `vlan` parameter. But when the bridge already exists, this module ignore the `vlan` parameter instead of updating the existing bridge.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

openvswitch_bridge

##### ANSIBLE VERSION

```paste below
2.8.0
```

##### STEPS TO REPRODUCE

```yaml
- name: create lxd bridges
  openvswitch_bridge:
    bridge: "fake-br"
    parent: "br0"
    vlan: "100"

- name: update lxd bridges
  openvswitch_bridge:
    bridge: "fake-br"
    parent: "br0"
    vlan: "200"
```

##### EXPECTED RESULTS

The first command create the bridge, the second one update the bridge and set the vlan to 200.

The bridge can be updated using the following command:

`ovs-vsctl set port fake-br tag=200`

##### ACTUAL RESULTS

`vlan` parameter is ignored and the command reports 'no change'

### PR DESCRIPTION
**Pull request by danielmellado**
##### SUMMARY
Add support for vlan update on ovs bridges

This commit adds support for updating vlan to an existing openvswitch bridge,
which was previously ignored-

Fixes #57018

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
openvswitch_brige

### CODE REVIEW

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 01772cd54b1fe0c132020adc5f3c8ae328bf583f

**Review by trishnaguha — APPROVED** [review]

**Comment by pabelanger** [comment]
We should check for ovs_vlan test, and maybe had some coverage here.

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 8cfdcdbe44eb84c3d1fd97636ef0e25ea92eda1f

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 6c6c97123a153da95da826b429a722acdb91e007

**Comment by pabelanger** [comment]
recheck

**Review by pabelanger — CHANGES_REQUESTED** [review]
-1, this is going to break out integration tests. If you can debug why that is.

File: lib/ansible/modules/network/ovs/openvswitch_bridge.py
Comment (trishnaguha) [inline lib/ansible/modules/network/ovs/openvswitch_bridge.py]: This is where it is failing
```
cmd: /usr/bin/ovs-vsctl -t 5 set port br-test tag=None
msg: 'ovs-vsctl: "None" is not a valid integer or range'
```
You will need a similar condition like `if want['vlan'] and want['vlan'] != have['vlan']:`

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 7caef2c497a1fe9c961bd9233f5a8713b51f61e2

**Comment by danielmellado** [comment]
Thanks @trishnaguha and @pabelanger, let's get this merged and I'll be adding integration and unit tests on a follow-up patch depending on this one!

**Comment by pabelanger** [comment]
I'd like to hold off on merging until we have new tests in place, given that we have zuul running integration testing, we can now validate this code works as expected before merging.

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 805d4ce9a82edf7fc18297367ee36716ef20a529

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 15c3dc6b817c5aa42f0434580754e00dade281ab

**Review by pabelanger — APPROVED** [review]
Thanks!

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 83bf4759e07bd7e17dcfad5b7f6451e766272a28

File: test/units/modules/network/ovs/test_openvswitch_bridge.py
Comment (webknjaz) [inline test/units/modules/network/ovs/test_openvswitch_bridge.py]: We usually prefer dict literals `{}` over `dict()` constructors/factories. (Although it's a holywar topic)

File: test/units/modules/network/ovs/test_openvswitch_bridge.py
Comment (webknjaz) [inline test/units/modules/network/ovs/test_openvswitch_bridge.py]: With pytest, we usually go for fixtures so that you'd have a separate fixture function patching staff and yielding some object for testing and the test function will be more readable and contain actual testing code with assertions.
For example:
```python
@pytest.fixture
def patched_openvswitch_bridge(monkeypatch):
    with monkeypatch.context() as mp_ctx:
        mp_ctx.setattr(
            openvswitch_bridge,
            'map_config_to_obj', 
            {
               'bridge': 'test-br2', 'parent': 'test-br',
               'vlan': 200, 'fail_mode': None,
               'external_ids': None, 'set': None,
            },
        )
        # code before "yield" is executed before test function
        yield openvswitch_bridge  # "yield"ed value is what test function receives as an input
        # code after "yield" is executed after test function
        # and can be used for cleanup, we don't need this because
        # it'll be taken care of by the context manager


def test_openvswitch_bridge_updates_vlan(self, patched_openvswitch_bridge):
    ...  # patched_openvswitch_bridge is patched inside this test method
```

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 838227205e02ab53807340eaf14011b88f913aa6

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 01435667e2b34c4efbff54076b393305647c33ab

File: test/units/modules/network/ovs/test_openvswitch_bridge.py
Comment (webknjaz) [inline test/units/modules/network/ovs/test_openvswitch_bridge.py]: Drop `with` block and patch things directly.
```suggestion
        monkeypatch.setattr(openvswitch_bridge, 'map_config_to_obj', mocked_bridge)
```

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 7ef7dd6396b48f890c233f9198f4fb58ed82c227

File: test/units/modules/network/ovs/test_openvswitch_bridge.py
Comment (webknjaz) [inline test/units/modules/network/ovs/test_openvswitch_bridge.py]: In this case, you don't need a generator, just
```suggestion
    return openvswitch_bridge
```

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now d80596089f01ccfd3685fe741b41e442bf41c0fc

File: lib/ansible/modules/network/ovs/openvswitch_bridge.py
Comment (webknjaz) [inline lib/ansible/modules/network/ovs/openvswitch_bridge.py:144]: I'm not sure what you check with `if want['vlan']` but it seems wrong without any explanation...

File: test/integration/targets/openvswitch_bridge/tests/basic.yaml
Comment (webknjaz) [inline test/integration/targets/openvswitch_bridge/tests/basic.yaml]: no need for quotes in this case..
```suggestion
      - result.changed == true
```

Comment (webknjaz) [reply test/integration/targets/openvswitch_bridge/tests/basic.yaml]: Oh, wait, also no need for the comparison with true: first of all, the operator should be `is` and second -- it's already truthy.
```suggestion
      - result.changed
```

File: test/integration/targets/openvswitch_bridge/tests/basic.yaml
Comment (webknjaz) [inline test/integration/targets/openvswitch_bridge/tests/basic.yaml]:
```suggestion
      - result.changed
```

Comment (danielmellado) [reply test/integration/targets/openvswitch_bridge/tests/basic.yaml]: All right, let's change it, although it is unrelated to the patch scope!

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 2a054d94d1e04ffacff1907fc5e704fbb4916148

File: test/units/modules/network/ovs/test_openvswitch_bridge.py
Comment (webknjaz) [inline test/units/modules/network/ovs/test_openvswitch_bridge.py:177]: Are you sure it doesn't work with just
```suggestion
    def test_openvswitch_bridge_updates_vlan(self, patched_openvswitch_bridge):
```
?

Comment (webknjaz) [reply test/units/modules/network/ovs/test_openvswitch_bridge.py:177]: Oh.... It derives from `unittest.TestCase` which is almost incompatible with pytest..

File: lib/ansible/modules/network/ovs/openvswitch_bridge.py
Comment (danielmellado) [reply lib/ansible/modules/network/ovs/openvswitch_bridge.py:144]: there's no explanation in the rest of the module either, so I think documenting the module is outside of the scope of this patch. In any case it basically checks that it's getting a vlan from args and such vlan is not the same as the one already existing on the bridge

File: test/integration/targets/openvswitch_bridge/tests/basic.yaml
Comment (samdoran) [inline test/integration/targets/openvswitch_bridge/tests/basic.yaml]: It's better to use the `changed` test. That looks through the results for a `changed` key.
```suggestion
      - result is changed
```

File: test/integration/targets/openvswitch_bridge/tests/basic.yaml
Comment (samdoran) [inline test/integration/targets/openvswitch_bridge/tests/basic.yaml]:
```suggestion
      - result is not changed
```

File: test/integration/targets/openvswitch_bridge/tests/basic.yaml
Comment (samdoran) [inline test/integration/targets/openvswitch_bridge/tests/basic.yaml]:
```suggestion
      - result is changed
```

File: test/integration/targets/openvswitch_bridge/tests/basic.yaml
Comment (samdoran) [inline test/integration/targets/openvswitch_bridge/tests/basic.yaml]:
```suggestion
      - result is changed
```

File: lib/ansible/modules/network/ovs/openvswitch_bridge.py
Comment (samdoran) [reply lib/ansible/modules/network/ovs/openvswitch_bridge.py:144]: The `if want['vlan']` test isn't needed. That value will default to `None` per the `argument_spec`.

**Review by samdoran — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/ovs/openvswitch_bridge.py
Comment (danielmellado) [reply lib/ansible/modules/network/ovs/openvswitch_bridge.py:144]: from a main pythonic perspective it won't be needed, but if we remove that part it might be conflicting with some test config fixtures where want['vlan'] might be None and have['vlan'] could be a fixture, such as 0. Having this check prevents a false check there.

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now aa7c9e89deeccb4e097604a666962a23198e9db2

**Review by webknjaz — APPROVED** [review]
Good to go!

**Comment by pabelanger** [comment]
There was a new release of ARA today (1.0.0) that broke our zuul jobs for network-integration, fixing now and will recheck shortly.

**Comment by pabelanger** [comment]
recheck

**Review by samdoran — APPROVED** [review]

**Comment by samdoran** [comment]
Please create a [backport PR](https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs) for this to be included in previous versions.

**Review by ganeshrn — APPROVED** [review]

**Pull request merged**
